### PR TITLE
Improve CCA news feature

### PR DESCRIPTION
**Pull request by benclmnt**
Move changes to branch from #33

### CODE REVIEW

**Comment by benclmnt** [comment]
Missing: how to load text file properly.

Looking into : txt file loader plugin for rollup. 
Question: Can we directly load txt file into server js routes?

File: src/routes/news/[...slug].svelte [hunk 1]
```diff
@@ -0,0 +1,152 @@
+<script context="module">
+    import newsTargets from 'news-targets';
+
+    export async function preload({ params, query }) {
+        // the `slug` parameter is available because
+        // this file is called [slug].svelte
```
Comment (Yehezkiel01) [inline src/routes/news/[...slug].svelte]: Outdated comments, we can either remove this or update it into [...slug].svelte

File: rollup.config.js [hunk 2]
```diff
@@ -16,11 +20,34 @@ const onwarn = (warning, onwarn) =>
         /[/\\]@sapper[/\\]/.test(warning.message)) ||
     onwarn(warning);
 
+// https://github.com/rollup/rollup/issues/2463#issuecomment-455957865
```
Comment (Yehezkiel01) [inline rollup.config.js:23]: The reference is clear but I believed it will be easier if we add some explanatory comments on what this function is trying to do. It takes me a while to understand why we can import 'news-targets'

File: src/components/Nav.svelte [hunk 3]
```diff
@@ -1,147 +1,147 @@
...
-        <li>
-            <a
-                aria-current={segment === 'contact' ? 'page' : undefined}
-                href="contact">
-                Contact Us
-            </a>
-        </li>
-    </ul>
-</nav>
+<script>
```
Comment (Yehezkiel01) [inline src/components/Nav.svelte:1]: I think there are some unnecessary whitespace here that mark the whole files to be changed. We can try to investigates this

File: src/routes/news/[...slug].svelte [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        const [slug, title] = params.slug;
```
Comment (Yehezkiel01) [inline src/routes/news/[...slug].svelte:6]: Why do we have slug and title in the news url? I believed slug is meant to be unique.

File: src/routes/news/[...slug].svelte [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+        const res = await this.fetch(`news/${slug}.json`);
+        const data = await res.json();
+
+        // if title exists, try to find news with matching title. if not found, get the first one
+        // TODO: current support for multiple articles with same date is still hacky
+        let news =
+            (title && data.news.find((news) => news.title == title)) ||
+            data.news[0];
+
+        // https://github.com/rollup/rollup/issues/2463#issuecomment-455957865
```
Comment (Yehezkiel01) [inline src/routes/news/[...slug].svelte:22]: Similar to my previous comment, a short description of what this function is trying to do will be helpful

File: src/routes/news/[...slug].svelte [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        news.file = await newsTargets[news.file]().then((x) => x.default);
+
+        if (res.status === 200) {
+            return {
+                news,
+                number: data.number,
+            };
+        } else {
+            this.error(res.status, data.message);
+        }
```
Comment (Yehezkiel01) [inline src/routes/news/[...slug].svelte]: I believed that this error handling could be moved above. Right after fetching res and data.

File: src/routes/news/[...slug].svelte [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    }
+</script>
+
+<script>
+    import { dates } from './_news.js';
+    export let number;
```
Comment (Yehezkiel01) [inline src/routes/news/[...slug].svelte]: I think number is not too clear. We can give better context by using 'newsIndex' instead?

File: src/routes/news/_news.js [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        'April',
+        'May',
+        'June',
+        'July',
+        'August',
+        'September',
+        'October',
+        'November',
+        'December',
+    ];
```
Comment (Yehezkiel01) [inline src/routes/news/_news.js]: It will be better to move this outside of the function. This will help keeping the function concise

File: src/routes/news/_news.js [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+        if (!news.date || !news.date.match(/[0-9]{4}-[0-9]{2}-[0-9]{2}/)) {
+            throw Error('Date format must be in YYYY-MM-DD format');
+        }
+
+        const [year, month, date] = news.date.split('-').map(Number);
+
+        // sanity date checking
+        if (month < 1 || month > 12 || date < 0 || date > 31) {
+            throw Error(`Invalid date ${news.date}`);
+        }
```
Comment (Yehezkiel01) [inline src/routes/news/_news.js:60]: Nice sanity check

File: src/routes/news/_news.js [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+        news.displayDate = `${date} ${monthList[month - 1]} ${year}`;
+        let datesWoSep = news.date.replace(/-/g, '');
```
Comment (Yehezkiel01) [inline src/routes/news/_news.js]: datesWoSep confused me initially, perhaps naming it 'datesWithoutSeparator' will make it much clearer. It's okay to use long name variables.

File: src/routes/news/_news.js [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    {
+        title: 'GHI',
+        file: '20201113.txt',
+        img: 'angklung.jpg',
+        img_name: 'angklung',
+        date: '2020-11-13',
+    },
+];
+
+const transform = (newsList) => {
```
Comment (Yehezkiel01) [inline src/routes/news/_news.js]: To make it clear, We can add some comment describing in what formats will this transform function returns

**Review by Yehezkiel01 — CHANGES_REQUESTED** [review]
The PR is working but I still commented some to increase our code quality ;)

File: src/routes/news/[...slug].svelte (hunk 4 above)
Comment (kevinchs0808) [reply src/routes/news/[...slug].svelte:6]: Noted. I've added a short description about slug in line 4-5

File: src/routes/news/[...slug].svelte (hunk 6 above)
Comment (kevinchs0808) [reply src/routes/news/[...slug].svelte]: No worries. I've moved the error handling statement to line 9-19, right after fetching res and data.

File: src/routes/news/_news.js [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    },
+    {
+        title: 'GHI',
+        file: '20201113.txt',
+        img: 'angklung.jpg',
+        img_name: 'angklung',
+        date: '2020-11-13',
+    },
+];
+
```
Comment (kevinchs0808) [inline src/routes/news/_news.js:27]: Moved the monthlist out of the transform function.

File: src/routes/news/_news.js [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+        'May',
+        'June',
+        'July',
+        'August',
+        'September',
+        'October',
+        'November',
+        'December',
+    ];
+
```
Comment (kevinchs0808) [inline src/routes/news/_news.js:42]: Added a short description about the transform function.

File: src/routes/news/[...slug].svelte [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+<br />
+<img
+    src={get_image_source(news.img)}
+    alt="{news.img_name} photo"
+    class="center_image" />
+<p>Date: {news.displayDate}</p>
+<br />
+<p>{news.file}</p>
+<br />
+<div class="navigation">
```
Comment (kevinchs0808) [inline src/routes/news/[...slug].svelte:80]: Changed the 'number' into 'newsIndex'

File: src/routes/news/[...slug].svelte [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+    const default_picture = 'client/news/default-picture.svg';
+    // convert image slug (if any) into image source or return a default source.
+    function get_image_source(image_slug) {
+        if (!image_slug) {
+            return default_picture;
+        }
+        return 'client/news/' + dates[number] + '/' + image_slug;
+    }
+</script>
+<style>
```
Comment (kevinchs0808) [inline src/routes/news/[...slug].svelte:43]: Added some descriptions for each CSS selectors and deleted several irrelevant CSS selectors

File: src/routes/news/_news.js [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+        }
+
+        const [year, month, date] = news.date.split('-').map(Number);
+
+        // sanity date checking
+        if (month < 1 || month > 12 || date < 0 || date > 31) {
+            throw Error(`Invalid date ${news.date}`);
+        }
+
+        news.displayDate = `${date} ${monthList[month - 1]} ${year}`;
```
Comment (kevinchs0808) [inline src/routes/news/_news.js:62]: Changed the datesWoSep into datesWithoutSeparator and updated an outdated comment (line 64)

File: src/components/Nav.svelte (hunk 3 above)
Comment (kevinchs0808) [reply src/components/Nav.svelte:1]: I haven't been able to notice all the whitespace, but the only whitespace that I can see is at the bottom line (the deleted line 148)

File: rollup.config.js (hunk 2 above)
Comment (kevinchs0808) [reply rollup.config.js:23]: I have added a short description about the 'newsTargetVirtualModule' (line 25-26). Feel free to edit the description, if there are any incorrect information.

**Review by kevinchs0808 — DISMISSED** [review]
Thank you for your guidances @Yehezkiel01 and @benclmnt 

Really appreciate your assistance.

I've made some changes:
-. Deleted unnecessary CSS selectors
-. Updated the outdated comments
-. Added short descriptions for several functions
-. Changed several variable names such as (number -> newsIndex) and (datesWoSep -> datesWithoutSeparator)

If there are any incorrect information, feel free to make any changes.

File: src/routes/news/[...slug].svelte [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+        const res = await this.fetch(`news/${slug}.json`);
+        const data = await res.json();
+        // This if-else statement is used to make sure that 
+        // the web page can be displayed (redirect status of 200). 
+        // Otherwise, an error message will be displayed.
+        if (res.status === 200) {
+            return {
+                news,
+                newsIndex: data.newsIndex,
+            };
```
Comment (Yehezkiel01) [inline src/routes/news/[...slug].svelte]: I think this is not supposed to be here since `news` is defined on line 23.
image

What I was suggesting before is that.
We simply moved the error-handling to this place.

So, we can have like
```
if (res.status != 200) {
    this.error(res.status, data.message);
    return; // we stop the function immediately
}
```

Later on after line 27, we just need to return the objects without any other if-else error checking since we have handled it above.

```
return {
    news,
    newsIndex: data.newsIndex,
};
```

**Review by kevinchs0808 — APPROVED** [review]

**Review by kevinchs0808 — APPROVED** [review]

**Review by Yehezkiel01 — APPROVED** [review]
LGTM! Thanks for updating the PR
